src/libos/src/signal/do_sigtimedwait.rs

```diff
@@ -26,13 +22,15 @@ pub async fn do_sigtimedwait(interest: SigSet, timeout: Option<&Duration>) -> Re
         *blocked & interest
     };
 
+    let mut timeout = timeout.cloned();
     // Loop until we find a pending signal or reach timeout
-    waiter_loop!(process.sig_waiters(), {
+    waiter_loop!(process.sig_waiters(), timeout, {
         if let Some(signal) = dequeue_pending_signal(&interest, &thread, &process) {
             let siginfo = signal.to_info();
             return Ok(siginfo);
         }
-    })
+    }) as Result<siginfo_t>;
```

Comment:
It is a bit of verbose to write `as Result<_>` and then `return_errno`. I think it can be simplified this way.

```rust
waiter_loop!(process.sig_waiters(), timeout, {
    //...
}).map_err(|e| errno!(EAGAIN, "no interesting, pending signal"))
```

In the future, when `wait` may return `EINTR`, then the code should be written like this.

```rust
waiter_loop!(process.sig_waiters(), timeout, {
    //...
}).map_err(|e| {
    if e.errno() == ETIMEDOUT { 
        errno!(EAGAIN, "no interesting, pending signal")
    } else {
        e
    }
})
```

The above version highlights the fact that `waiter_loop` always returns a `Result<_>` and the user can deal with it (instead of ignoring it).

Reply:
According to linux doc, the errno should ETIMEDOUT

> ETIMEDOUT
>               The operation in futex_op employed the timeout specified
>               in timeout, and the timeout expired before the operation
>               completed.

> EAGAIN (FUTEX_WAIT, FUTEX_WAIT_BITSET, FUTEX_WAIT_REQUEUE_PI) The
>               value pointed to by uaddr was not equal to the expected
>               value val at the time of the call.

```
waiter_loop!(process.sig_waiters(), timeout, {
    //...
}).map_err(|e| {
    if e.errno() == ETIME { 
        errno!(ETIMEDOUT, "no interesting, pending signal")
    } else {
        e
    }
})
```